src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.Impl.cs

```diff
@@ -293,21 +295,23 @@ private void OnSymbolActionCore(ISymbol symbol, Action<Diagnostic> reportDiagnos
                         method.ContainingType.TypeKind == TypeKind.Class &&
                         !method.ContainingType.IsSealed &&
                         method.ContainingType.BaseType != null &&
-                        IsPublicApiCore(method.ContainingType.BaseType) &&
-                        !CanTypeBeExtendedPublicly(method.ContainingType.BaseType))
+                        IsTrackedApiCore(method.ContainingType.BaseType) &&
+                        !CanTypeBeExtended(method.ContainingType.BaseType))
                     {
                         string errorMessageName = GetErrorMessageName(method, isImplicitlyDeclaredConstructor);
                         ImmutableDictionary<string, string> propertyBag = ImmutableDictionary<string, string>.Empty;
                         var locations = isImplicitlyDeclaredConstructor ? method.ContainingType.Locations : method.Locations;
-                        reportDiagnostic(Diagnostic.Create(ExposedNoninstantiableType, locations[0], propertyBag, errorMessageName));
+                        reportDiagnostic(Diagnostic.Create(GetDiagnostic(ExposedNoninstantiableTypePublic, ExposedNoninstantiableTypeInternal), locations[0], propertyBag, errorMessageName));
                     }
 
                     // Flag public API with optional parameters that violate backcompat requirements: https://github.com/dotnet/roslyn/blob/main/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md.
                     if (method.HasOptionalParameters())
                     {
                         foreach (var overload in method.GetOverloads())
                         {
-                            if (!IsPublicAPI(overload))
+                            var symbolAccessibility = overload.GetResultantVisibility();
+                            var minAccessibility = _isPublic ? SymbolVisibility.Public : SymbolVisibility.Internal;
+                            if (symbolAccessibility > minAccessibility)
```

Comment:
Can't just use `IsTrackedApi` here anymore, as the internal APIs can be affected by public overloads with optional parameters.

Reply:
I left a comment in the tests where this is observable.